### PR TITLE
Generate cache Authfile

### PR DESCRIPTION
**Pull request by bbockelm**
This performs matchmaking between origins and caches before adding a given origin to the cache's authfile.

This way, for example, caches can refuse to host data from a given origin server.

Tests well via CLI but is blocked by #335.  Will rebase once that is merged.

### CODE REVIEW

**Comment by matyasselmeci** [comment]
#335 is merged. Can you rebase?

**Comment by matyasselmeci** [comment]
@brianhlin this is ready for review. There are a couple of items that should be done but shouldn't hold this PR up:
- do a more thorough test of Authfile generation in the CI instead of just trying one host
- factor out some duplicated code in `stashcache.py` (possibly moving some of it to `topology.py`)

**Comment by matyasselmeci** [comment]
Before we merge this -- what does "ANY" mean for AllowedVOs? How is it different from "PUBLIC"?

**Comment by bbockelm** [comment]
What do you mean by "authenticated" as a VO?  ANY implies that the cache is willing to serve any OSG VO that is willing to utilize the cache (i.e., it is a wildcard): PUBLIC implies that the cache is willing to host any public data.

File: virtual-organizations/CMS.yaml [hunk 1]
```diff
@@ -117,3 +117,4 @@ DataFederations:
     Namespaces:
       /store:
         - FQAN:/cms
+    AllowedCaches: []
```
Comment (brianhlin) [inline virtual-organizations/CMS.yaml]: I'm not sure I understand this -- doesn't this prevent CMS from being in any automatically generated cache authfiles?

**Comment by matyasselmeci** [comment]
> What do you mean by "authenticated" as a VO? ANY implies that the cache is willing to serve any OSG VO that is willing to utilize the cache (i.e., it is a wildcard): PUBLIC implies that the cache is willing to host any public data.

So "ANY" is for output and "PUBLIC" is for input?

**Comment by matyasselmeci** [comment]
> A few minor comments/questions but otherwise it looks reasonable. It also shouldn't hold up the PR but it seems silly to have separate `osg-authfile` and `osg-authfile-public`.

There are two instances of xrootd running and one of them uses osg-authfile and the other uses osg-authfile-public.

**Comment by brianhlin** [comment]
> So "ANY" is for output and "PUBLIC" is for input?

I'm not really sure what you mean here but it doesn't seem right. I believe both of them suggest that the cache is VO-agnostic, the former for private data and the latter for public data.

> There are two instances of xrootd running and one of them uses osg-authfile and the other uses osg-authfile-public.

Couldn't we just have the public version be an option, e.g. `osg-authfile --public`?

File: src/app.py [hunk 2]
```diff
@@ -179,6 +161,31 @@ def origin_authfile():
...
+                                 legacy=app.config["STASHCACHE_LEGACY_AUTH"],
+                                 suppress_errors=False)
+    except stashcache.DataError as e:
+        app.logger.error("{}: {}".format(request.full_path, str(e)))
+        return Response("# Error generating authfile for this FQDN: {}\n".format(str(e)) +
+                        "# Please check configuration in OSG topology or contact support@opensciencegrid.org\n",
+                        mimetype="text/plain", status=400)
+    except Exception:
+        app.log_exception(sys.exc_info())
+        return Response("Server error getting authfile", status=503)
```
[the remark this answers is not in the document]
Comment (brianhlin) [reply src/app.py]: Sure when we have that set up we could change the message to

> Server error getting authfile, the administrator has been notified

But until we have that, we should say:

```suggestion
        return Response("Server error getting authfile, please contact help@opensciencegrid.org", status=503)
```

**Comment by matyasselmeci** [comment]
> > So "ANY" is for output and "PUBLIC" is for input?
> 
> I'm not really sure what you mean here but it doesn't seem right. I believe both of them suggest that the cache is VO-agnostic, the former for private data and the latter for public data.

The fact that we are having this discussion indicates that we need to change something in order for this to be clear.
 
> > There are two instances of xrootd running and one of them uses osg-authfile and the other uses osg-authfile-public.
> 
> Couldn't we just have the public version be an option, e.g. `osg-authfile --public`?

We could. The `osg-authfile` script is for internal/CI use anyway; the actual origins/caches will periodically use `wget` to fetch the information from the topology webapp.

File: src/app.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (matyasselmeci) [reply src/app.py]: That change should be in a separate PR with corresponding changes throughout the rest of  the webapp.

[the remark this answers is not in the document]
Comment (matyasselmeci) [reply src/app.py]: I can update it here. What I meant was to create another PR for making the same change everywhere else in the webapp.

File: virtual-organizations/CMS.yaml (hunk 1 above)
Comment (matyasselmeci) [reply virtual-organizations/CMS.yaml]: @bbockelm ^^

File: topology/University of Wisconsin/CHTC/CHTC.yaml [hunk 3]
```diff
@@ -266,6 +266,8 @@ Resources:
         Description: StashCache cache server
     VOOwnership:
       GLOW: 100
+    AllowedVOs:
+      - PUBLIC
```
[the remark this answers is not in the document]
Comment (matyasselmeci) [reply topology/University of Wisconsin/CHTC/CHTC.yaml]: I changed it.

File: src/app.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (matyasselmeci) [reply src/app.py]: Done.

File: src/app.py [hunk 4]
```diff
@@ -179,6 +161,31 @@ def origin_authfile():
...
+            generate_function = stashcache.generate_authfile
+        auth = generate_function(global_data.get_vos_data(),
+                                 global_data.get_topology().get_resource_group_list(),
+                                 fqdn=cache_fqdn,
+                                 legacy=app.config["STASHCACHE_LEGACY_AUTH"],
+                                 suppress_errors=False)
+    except stashcache.DataError as e:
+        app.logger.error("{}: {}".format(request.full_path, str(e)))
+        return Response("# Error generating authfile for this FQDN: {}\n".format(str(e)) +
+                        "# Please check configuration in OSG topology or contact support@opensciencegrid.org\n",
```
[the remark this answers is not in the document]
Comment (matyasselmeci) [reply src/app.py]: Done.

**Pull request merged**